File: SwiftUI/WatchWhat/Shared/GraphQL/Movie.swift
//
//  Movie.swift
//  WatchWhat
//
//  Created by Drew Miller on 3/8/23.
//

import Foundation

struct Movie {
    let adult: Bool
    let backdropPath: String?
    let genreIds: [Int]
    let id: Int
    let originalLanguage: String
    let originalTitle: String
    let overview: String
    let popularity: Double
    let posterPath: String?
    
    private let releaseDateString: String
    let var releaseDate: Date {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.timeZone = TimeZone.current
        dateFormatter.locale = Locale.current
        return dateFormatter.date(from: self.releaseDateString)
    }
    
    let title: String
    let video: Bool
    let voteAverage: Bool
    let voteAverage: Double
    let voteCount: Int
    
    private init() {}
    
    func static fromPopularMovies(movie: PopularMoviesResult): Movie {
        self.adult              = movie.adult
        self.backdropPath       = movie.backdrop_path
        self.genreIds           = movie.genre_ids
        self.id                 = movie.id
        self.originalLanguage   = movie.original_language
        self.originalTitle      = movie.original_title
        self.popularity         = movie.popularity
        self.posterPath         = movie.poster_path
        self.releaseDate        = movie.release_date // Convert
        self.title              = movie.title
        self.video              = movie.video
        self.voteAverage        = movie.vote_average
        self.voteCount          = movie.vote_count
    }
}

File: Function.GraphQL/Shared/tmdb/dtos/movie.ts
import { Genre } from "./genre";

export type Movie = {
  adult: boolean,
  backdrop_path: string,
  genre_ids: number[],
  id: number,
  original_language: string,
  original_title: string,
  overview: string,
  popularity: number,
  poster_path: string,
  release_date: Date,
  title: string,
  video: boolean,
  vote_average: number,
  vote_count: number
};

// This property is likely a collection group.
// belongs_to_collection: any | null,
export type MovieDetails = {
  adult: boolean,
  backdrop_path: string,
  budget: number,
  id: number,
  genres: Genre[],
  homepage: string,
  imdb_id: string,
  original_language: string,
  original_title: string,
  overview: string,
  popularity: number,
  poster_path: string,
  production_companies: ProductionCompany[],
  production_countries: ProductionCountry[],
  release_date: Date,
  revenue: number,
  runtime: number,
  spoken_languages: SpokenLanguages[],
  status: string,
  tagline: string,
  title: string,
  video: boolean,
  vote_average: number,
  vote_count: number
};

type ProductionCompany = {
  id: number,
  logo_path: string,
  name: string,
  origin_country: string
};

type ProductionCountry = {
  iso_3166_1: string,
  name: string
}

type SpokenLanguages = {
  english_name: string,
  iso_639_1: string,
  name: string
}